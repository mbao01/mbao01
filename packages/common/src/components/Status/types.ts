import { type VariantProps } from "../../libs";
import { getStatusClasses } from "./constants";

export type StatusProps = Omit<React.HTMLAttributes<HTMLSpanElement>, "children"> &
  VariantProps<typeof getStatusClasses>;
