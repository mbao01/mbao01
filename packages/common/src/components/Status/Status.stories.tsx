import type { Meta, StoryObj } from "@storybook/react";
import { Status } from "./Status";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Status,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Status>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};

export const TinyStatus: Story = {
  args: {
    size: "xs",
  },
};

export const InfoStatus: Story = {
  args: {
    animate: "ping",
    variant: "info",
  },
  render: (args) => {
    return (
      <>
        <Status {...args} /> New message
      </>
    );
  },
};

export const PrimaryStatus: Story = {
  args: {
    variant: "primary",
  },
};

export const SuccessStatus: Story = {
  args: {
    variant: "success",
  },
};
