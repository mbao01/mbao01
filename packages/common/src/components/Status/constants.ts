import { cva } from "../../libs";

export const getStatusClasses = cva("status", {
  variants: {
    variant: {
      accent: "status-accent",
      default: "",
      error: "status-error",
      ghost: "",
      info: "status-info",
      neutral: "status-neutral",
      primary: "status-primary",
      secondary: "status-secondary",
      success: "status-success",
      warning: "status-warning",
    },
    size: {
      xs: "status-xs",
      sm: "status-sm",
      md: "status-md",
      lg: "status-lg",
      xl: "status-xl",
    },
    animate: {
      bounce: "animate-bounce",
      ping: "animate-ping",
      pulse: "animate-pulse",
      spin: "animate-spin",
    },
  },
});
