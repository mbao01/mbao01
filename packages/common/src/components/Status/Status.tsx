import type { StatusProps } from "./types";
import { cn } from "../../utilities";
import { getStatusClasses } from "./constants";

export const Status = ({ animate, variant, size, className, ...props }: StatusProps) => (
  <span className={cn(getStatusClasses({ animate, variant, size }), className)} {...props} />
);
