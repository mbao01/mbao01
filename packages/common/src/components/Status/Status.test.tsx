import { render, screen } from "@testing-library/react";
import { Status } from "./Status";

describe("Status", () => {
  it("renders a basic status", () => {
    const { asFragment } = render(<Status />);

    expect(screen.getByText("K")).toBeVisible();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    { size: "xs", name: "tiny" },
    { size: "sm", name: "small" },
    { size: "md", name: "base" },
    { size: "lg", name: "large" },
  ] as const)("has $name ($size) button", ({ size, name }) => {
    const title = `Status ${name}`;
    const { asFragment } = render(
      <span data-testid="content">
        {title}: <Status size={size} />
      </span>
    );

    expect(screen.getByTestId("content")).toHaveTextContent(title);
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    "neutral",
    "primary",
    "secondary",
    "accent",
    "ghost",
    "info",
    "success",
    "warning",
    "error",
  ] as const)("has %s variant", (variant) => {
    const title = `Status ${variant}`;
    const { asFragment } = render(
      <span data-testid="content">
        {title}: <Status variant={variant} />
      </span>
    );

    expect(screen.getByTestId("content")).toHaveTextContent(`${title}: B`);
    expect(asFragment()).toMatchSnapshot();
  });
});
