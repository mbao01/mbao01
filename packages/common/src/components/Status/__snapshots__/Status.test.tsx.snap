// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Status > has 'base' ('md') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status base: 
    <span
      class="status status-md"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has 'large' ('lg') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status large: 
    <span
      class="status status-lg"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has 'small' ('sm') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status small: 
    <span
      class="status status-sm"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has 'tiny' ('xs') button 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status tiny: 
    <span
      class="status status-xs"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has accent variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status accent: 
    <span
      class="status status-accent"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has error variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status error: 
    <span
      class="status status-error"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has ghost variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status ghost: 
    <span
      class="status"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has info variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status info: 
    <span
      class="status status-info"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has neutral variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status neutral: 
    <span
      class="status status-neutral"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has primary variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status primary: 
    <span
      class="status status-primary"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has secondary variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status secondary: 
    <span
      class="status status-secondary"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has success variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status success: 
    <span
      class="status status-success"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > has warning variant 1`] = `
<DocumentFragment>
  <span
    data-testid="content"
  >
    Status warning: 
    <span
      class="status status-warning"
    />
  </span>
</DocumentFragment>
`;

exports[`Status > renders a basic status 1`] = `
<DocumentFragment>
  <span
    class="status"
  />
</DocumentFragment>
`;
